master/build-rust-manifest.py

```diff
@@ -0,0 +1,489 @@
+        cargo_date = most_recent_build_date("nightly", "cargo")
+    else:
+        # Beta / stable rust is paired with a cargo specified
+        # by rust-packaging
+        cargo_date = cargo_date_from_packaging(rustc_short_version)
+
+    cargo_version = version_from_channel("nightly", "cargo", cargo_date)
+    cargo_short_version = parse_short_version(cargo_version)
+
+    print "rustc date: " + rustc_date
+    print "rustc version: " + rustc_version
+    print "rustc short version: " + rustc_short_version
+    print "cargo date: " + cargo_date
+    print "cargo version: " + cargo_version
+    print "cargo short version: " + cargo_short_version
+
+    # Validate the component artifacts and generate the manifest
+    generate_manifest(rustc_date, rustc_version, rustc_short_version,
+                      cargo_date, cargo_version, cargo_short_version)
+
+# Use the channel-$component-$channel-date.txt file to get the archive
+# date for cargo or rustc. These files are generated by finish_dist
+# in rust-buildbot.
+def most_recent_build_date(channel, component):
+    dist = dist_folder(component)
+    date_url = s3_addy + "/" + dist + "/channel-" + component + "-" + channel + "-date.txt"
+    print "downloading " + date_url
+    response = urllib2.urlopen(date_url)
+    if response.getcode() != 200:
+        raise Exception("couldn't download " + date_url)
+    date = response.read().strip();
+    return date
+
+# Read the v1 manifests to find the installer name, download it
+# and extract the version file.
+def version_from_channel(channel, component, date):
+    dist = dist_folder(component)
+
+    # Load the manifest
+    manifest_url = s3_addy + "/" + dist + "/" + date + "/channel-" + component + "-" + channel
+    print "downloading " + manifest_url
+    response = urllib2.urlopen(manifest_url)
+    if response.getcode() != 200:
+        raise Exception("couldn't download " + manifest_url)
+    manifest = response.read().strip();
+
+    # Find the installer name
+    installer_name = None
+    for line in manifest.split("\n"):
+        if line.startswith(component) and line.endswith(".tar.gz"):
+            installer_name = line
+
+    if installer_name == None:
+        raise Exception("couldn't find installer in manifest for " + component)
+
+    # Download the installer
+    installer_url = s3_addy + "/" + dist + "/" + date + "/" + installer_name
+    print "downloading " + installer_url
+    response = urllib2.urlopen(installer_url)
+    if response.getcode() != 200:
+        raise Exception("couldn't download " + installer_url)
+
+    installer_file = temp_dir + "/" + installer_name
+    f = open(installer_file, "w")
+    while True:
+        buf = response.read(4096)
+        if not buf: break
+        f.write(buf)
+    f.close()
+
+    # Unpack the installer
+    unpack_dir = temp_dir + "/unpack"
+    os.mkdir(unpack_dir)
+    r = subprocess.call(["tar", "xzf", installer_file, "-C", unpack_dir, "--strip-components", "1"])
```

Comment:
I believe python has a built in `tarfile` library which can manage all this without frobbing the filesystem and shelling out, maybe not worth using though if this is already tested